torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2358,6 +2382,7 @@ def state_dict(self, *args, **kwargs):
         if torch.cuda.is_available():
             torch.cuda.synchronize()
         self._lazy_init()
+        self._clear_grads_if_needed()
```

Comment:
why do we need to do this for state_dict? Grads being none shouldn't matter there?

Reply:
I am just using the major FSDP calls as an entry point to release gradient memory as early as possible. In the code crawl I did manually, I found that sometimes people will checkpoint after `zero_grad(set_to_none=True)` after the optimizer step.